src/su/context.rs

```diff
@@ -23,15 +27,21 @@ impl SuContext {
         let process = crate::system::Process::new();
 
         // resolve environment, reset if this is a login
-        let mut environment = if options.login {
+        let mut environment = if options.login || !options.preserve_environment {
```

Comment:
I don't think this is correct. `su` does *not* clear the environment when `--preserve-environment`. `man su` says:

>        -m, -p, --preserve-environment
>           Preserve the entire environment, i.e., do not set HOME, SHELL, USER or LOGNAME. This option is ignored if the option --login is specified.

which I think means that only `HOME`, `SHELL`, `USER` and `LOGNAME` are not overwritten when the flag is used.

I think is reason why the `su::env::vars_in_invoking_users_env_are_preserved` test no longer passes with this PR.

Reply:
created a new issue for this #529